### PR TITLE
Fix for review's comment

### PR DESCRIPTION
**Pull request by kito-cheng**

### CODE REVIEW

File: gcc/config/riscv/riscv.c [hunk 1]
```diff
@@ -3417,6 +3412,44 @@ riscv_first_stack_step (struct riscv_frame_info *frame)
...
+      {
+	/* The save order is ra, s0, s1, s2 to s11.  */
+	if (regno == RETURN_ADDR_REGNUM)
+	  offset = saved_size - UNITS_PER_WORD;
+	else if (regno == S0_REGNUM)
+	  offset = saved_size - UNITS_PER_WORD * 2;
+	else if (regno == S1_REGNUM)
+	  offset = saved_size - UNITS_PER_WORD * 3;
+	else
+	  offset = saved_size - ((regno - S2_REGNUM) * UNITS_PER_WORD) - 16;
```
Comment (aswaterman) [inline gcc/config/riscv/riscv.c]: This looks wrong for RV64; instead of ```- 16``` should it be ```regno - S2_REGNUM + 4```?

**Comment by aswaterman** [comment]
Can you run through regressions with -msave-restore? 

    diff --git a/baseboards/riscv-sim.exp b/baseboards/riscv-sim.exp
    index 799e1cb..937b60c 100644
    --- a/baseboards/riscv-sim.exp
    +++ b/baseboards/riscv-sim.exp
    @@ -36,7 +36,7 @@ set_board_info compiler "[find_gcc]"
     
     # The basic set of flags needed to build "hello world" for this
     # board. This board uses libgloss and newlib.
    -set_board_info cflags 	"[libgloss_include_flags] [newlib_include_flags]"
    +set_board_info cflags 	"[libgloss_include_flags] [newlib_include_flags] -msave-restore"
     set_board_info ldflags 	"[libgloss_link_flags] [newlib_link_flags]"
     
     # This board doesn't use a linker script.

File: gcc/config/riscv/riscv.c (hunk 1 above)
Comment (kito-cheng) [inline gcc/config/riscv/riscv.c]: Oh, you are right...

**Comment by kito-cheng** [comment]
Test result is same as without -msave-restore exclude sibcall-9.c, sibcall-10.c became unsupported, I am also revert `Remove CALL_REALLY_USED_REGISTERS since it's same as CALL_USED_REGISTERS` this commit since its seem break the code gen and make the program hang even a simple hello world...this is strange I will investigate what happen today.

```
                                    |     rv32ima   |   rv32imafd   |     rv64ima   |   rv64imafd   |    rv32imaf   |    rv64imaf   | 
gcc       # of expected failures    |   173 (    0) |   173 (    0) |   174 (    0) |   174 (    0) |   173 (    0) |   174 (    0) | 
          # of expected passes      | 89833 (   -1) | 89937 (   -4) | 90822 (   -4) | 90934 (   -4) | 89938 (   -4) | 90931 (   -4) | 
          # of unexpected failures  |    16 (   -3) |    29 (    0) |    23 (    0) |    32 (    0) |    27 (    0) |    30 (    0) | 
          # of unexpected successes |     2 (    0) |     2 (    0) |     2 (    0) |     2 (    0) |     2 (    0) |     2 (    0) | 
          # of unsupported tests    |  2055 (    2) |  1997 (    2) |  1863 (    2) |  1803 (    2) |  1997 (    2) |  1805 (    2) | 
g++       # of expected failures    |   323 (    0) |   323 (    0) |   323 (    0) |   323 (    0) |   323 (    0) |   323 (    0) | 
          # of expected passes      | 97420 (    0) | 97420 (    0) | 99302 (    0) | 99302 (    0) | 97420 (    0) | 99302 (    0) | 
          # of unexpected failures  |    12 (    0) |    12 (    0) |    18 (    0) |    18 (    0) |    12 (    0) |    18 (    0) | 
          # of unexpected successes |     2 (    0) |     2 (    0) |     2 (    0) |     2 (    0) |     2 (    0) |     2 (    0) | 
          # of unsupported tests    |  4274 (    0) |  4274 (    0) |  4243 (    0) |  4243 (    0) |  4274 (    0) |  4243 (    0) | 
```
Note: base is test result without -msave-restore, # of unexpected failures -3 for rv32ima is the random noise which came from time-profiler-2.c

**Comment by aswaterman** [comment]
OK, I'll merge once I finish with the TRULY_NOOP_TRUNCATION stuff (it touches like 1000 lines of code...).

**Comment by aswaterman** [comment]
I've merged this locally, and will push once I fix my TRULY_NOOP_TRUNCATION regressions. Thanks!!

**Comment by kito-cheng** [comment]
Looking forward to your patch :)

**Pull request closed without merging**